### ISSUE TITLE
Create repositories for BsonDocument or dynamic types

### ISSUE THREAD

**Issue by pedrompc** [johnknoop/MongoRepository#17]
Hello,

I'm experimenting with your library and I was wondering if you have any advice regarding the usage of your library with more flexible types, such as BsonDocument or dynamic. 
As I understand it is currently not possible, as all types need to be previously registered with the Configure method, and there doesn't seem to be a way to create a repository by collection name, only by type parameter.
However, I have some cases in which I'm using the MongoDb driver and the queries are done using the aforementioned types, as the documents don't have a strictly defined schema. 
Given this limitation, I am still interested in using your library due to the TransactionalScope support it provides.
I think, in short, I wish that something like this was possible:

`client.GetRepository<dynamic>(collectionName, tenantName)`

I'd like to know your thought on this, and if you have any advice as to how I could achieve what I've described.

Thank you, and keep up the good work.

**Comment by johnknoop** [comment]
Hi.

As you point out, it's not currently supported (all collections need to be mapped upfront, so that the repo can create indices/capped collections etc), but it doesn't sound like an unreasonable feature request.

I'll have a look within the upcoming days to see if it would be easy to add.

**Comment by johnknoop** [comment]
I've done some experimenting allowing creation of repository instances using dynamic without mapping the collection upfront, and there are some issues with it.

I have to user `dynamic` since all of the methods of IRepository<T> expects objects of type T, so BsonDocuments won't work.

But issue with using dynamic in this case is that all update operations of IRepository<T> take `Func<UpdateDefinitionBuilder<TEntity>, UpdateDefinition<TEntity>>` instead of just `UpdateDefinition<TEntity>` This is to allow for syntax like this:

```cs
await repo.UpdateOneAsync(
    filter: x => x.SomeProp == "SomeValue"),
    update: x => x.Set(y => y.SomeOtherProp, "newvalue");
```
instead of the more verbose
```cs
await repo.UpdateOneAsync(
    filter: x => x.SomeProp == "SomeValue"),
    update: Builders<MyType>.Update.Set(y => y.SomeOtherProp, "newvalue");
```

And since expression trees may not contain dynamics, I thought I'd try something like this:

```cs
await repo.UpdateOneAsync(
    filter: Builders<MyType>.Filter.Eq("MyField", 1),
    update: _ => Builders<Dummy>.Update.Set("Name", "newname").ToBsonDocument());
```

But that generates invalid updates.

So in short: there's no simple clear-cut way to do this quickly. My recommendation would be that if you don't really need the majority of the functionality that this library provides, and just want to use the support for ambient transactions, then I'd say just implement that bit on your own. It's not particularly difficult, you can have a look in the code file MongoRepository.cs and see how I did it.

**Comment by pedrompc** [comment]
Hello,

Thanks very much for your feedback. That sounds like a good idea, I'll follow your suggestion and try to implement ambient transaction support on my side, using your code as a baseline.
Thank you for your help.
